src/core/core.tooltip.js

```diff
@@ -96,853 +96,859 @@ defaults._set('global', {
 
-					me.update(true);
-					me.pivot();
-				}
-			}
+			// Body
+			this.drawBody(pt, vm, ctx, opacity);
 
-			return changed;
+			// Footer
+			this.drawFooter(pt, vm, ctx, opacity);
 		}
-	});
+	},
 
 	/**
-	 * @namespace Chart.Tooltip.positioners
+	 * Handle an event
+	 * @private
+	 * @param {IEvent} event - The event to handle
+	 * @returns {Boolean} true if the tooltip changed
 	 */
-	Chart.Tooltip.positioners = {
-		/**
-		 * Average mode places the tooltip at the average position of the elements shown
-		 * @function Chart.Tooltip.positioners.average
-		 * @param elements {ChartElement[]} the elements being displayed in the tooltip
-		 * @returns {Point} tooltip position
-		 */
-		average: function(elements) {
-			if (!elements.length) {
-				return false;
-			}
+	handleEvent: function(e) {
+		var me = this;
+		var options = me._options;
+		var changed = false;
 
-			var i, len;
-			var x = 0;
-			var y = 0;
-			var count = 0;
-
-			for (i = 0, len = elements.length; i < len; ++i) {
-				var el = elements[i];
-				if (el && el.hasValue()) {
-					var pos = el.tooltipPosition();
-					x += pos.x;
-					y += pos.y;
-					++count;
-				}
-			}
+		me._lastActive = me._lastActive || [];
 
-			return {
-				x: Math.round(x / count),
-				y: Math.round(y / count)
-			};
-		},
-
-		/**
-		 * Gets the tooltip position nearest of the item nearest to the event position
-		 * @function Chart.Tooltip.positioners.nearest
-		 * @param elements {Chart.Element[]} the tooltip elements
-		 * @param eventPosition {Point} the position of the event in canvas coordinates
-		 * @returns {Point} the tooltip position
-		 */
-		nearest: function(elements, eventPosition) {
-			var x = eventPosition.x;
-			var y = eventPosition.y;
-			var minDistance = Number.POSITIVE_INFINITY;
-			var i, len, nearestElement;
-
-			for (i = 0, len = elements.length; i < len; ++i) {
-				var el = elements[i];
-				if (el && el.hasValue()) {
-					var center = el.getCenterPoint();
-					var d = helpers.distanceBetweenPoints(eventPosition, center);
-
-					if (d < minDistance) {
-						minDistance = d;
-						nearestElement = el;
-					}
-				}
-			}
+		// Find Active Elements for tooltips
+		if (e.type === 'mouseout') {
+			me._active = [];
+		} else {
+			me._active = me._chart.getElementsAtEventForMode(e, options.mode, options);
+		}
 
-			if (nearestElement) {
-				var tp = nearestElement.tooltipPosition();
-				x = tp.x;
-				y = tp.y;
-			}
+		// Remember Last Actives
+		changed = !helpers.arrayEquals(me._active, me._lastActive);
 
-			return {
-				x: x,
-				y: y
-			};
+		// Only handle target event on tooltip change
+		if (changed) {
+			me._lastActive = me._active;
+
+			if (options.enabled || options.custom) {
+				me._eventPosition = {
+					x: e.x,
+					y: e.y
+				};
+
+				me.update(true);
+				me.pivot();
+			}
 		}
-	};
-};
+
+		return changed;
+	}
+});
+
+/**
+ * @namespace Chart.Tooltip.positioners
+ */
+exports.positioners = positioners;
```

Comment:
@etimberg I'm not sure why we need to expose positioners (`Chart.Tooltip.positioners`)

Reply:
It must be a backwards compatibility thing. Not sure when it first got introduced

Reply:
Actually, I think it was public to allow users to implement their own logic:

```javascript
Chart.Tooltip.positioners.foobar = function(elements) {
  // ...
};

options: {
  tooltips: {
    position: 'foobar'
  }
}
```

Reply:
http://www.chartjs.org/docs/latest/configuration/tooltip.html#position-modes